velox/dwio/dwrf/reader/DwrfData.cpp

```diff
@@ -20,15 +20,19 @@
 
 namespace facebook::velox::dwrf {
 
-DwrfData::DwrfData(
-    std::shared_ptr<const dwio::common::TypeWithId> nodeType,
-    StripeStreams& stripe,
-    FlatMapContext flatMapContext)
-    : memoryPool_(stripe.getMemoryPool()),
-      nodeType_(std::move(nodeType)),
-      flatMapContext_(std::move(flatMapContext)),
-      rowsPerRowGroup_{stripe.rowsPerRowGroup()} {
+void DwrfData::init(StripeStreams& stripe) {
+  auto format = stripe.format();
+  if (format == DwrfFormat::kDwrf) {
+    initDwrf(stripe);
+  } else {
+    VELOX_CHECK(format == DwrfFormat::kOrc);
+    initOrc(stripe);
+  }
+}
+
+void DwrfData::initDwrf(StripeStreams& stripe) {
```

Comment:
How about we use a little template magic to reduce code duplication?

```c++
void DwrfData::init(StripeStreams& stripe) {
  auto format = stripe.format();
  if (format == DwrfFormat::kDwrf) {
    init<DwrfFormat::kDwrf>(stripe);
  } else {
    VELOX_CHECK(format == DwrfFormat::kOrc);
    init<DwrfFormat::kOrc>(stripe);
  }
}

template <DwrfFormat kFormat>
void DwrfData::init(StripeStreams& stripe) {
  EncodingKey encodingKey{nodeType_->id, flatMapContext_.sequence};

  auto kind = kFormat == DwrfFormat::kDwrf ? proto::Stream_Kind_PRESENT : proto::orc::Stream_Kind_PRESENT;
  std::unique_ptr<dwio::common::SeekableInputStream> stream = stripe.getStream(
      encodingKey.forKind(kind), false);
  if (stream) {
    notNullDecoder_ = createBooleanRleDecoder(std::move(stream), encodingKey);
  }

  // We always initialize indexStream_ because indices are needed as
  // soon as there is a single filter that can trigger row group skips
  // anywhere in the reader tree. This is not known at construct time
  // because the first filter can come from a hash join or other run
  // time pushdown.
  kind = kFormat == DwrfFormat::kDwrf ? proto::Stream_Kind_ROW_INDEX : proto::orc::Stream_Kind_ROW_INDEX;
  indexStream_ = stripe.getStream(
      encodingKey.forKind(proto::orc::Stream_Kind_ROW_INDEX), false);
}
```

Reply:
sounds good

Reply:
the operator `condition ? a : b` require that type a is the same as b, while `proto::Stream_Kind_PRESENT` is a different enum type with `proto::orc::Stream_Kind_PRESENT`, so it does not work.

so, let's keep it this way, a method for `dwrf` and a method for `orc`.

Reply:
We should avoid these duplications as it would increase maintenance burden.  We can do it like this instead:

```c++
template <DwrfFormat kFormat>
void DwrfData::init(StripeStreams& stripe) {
  EncodingKey encodingKey{nodeType_->id, flatMapContext_.sequence};
  auto id = kFormat == DwrfFormat::kDwrf ? encodingKey.forKind(proto::Stream_Kind_PRESENT) : encodingKey.forKind(proto::orc::Stream_Kind_PRESENT);
  std::unique_ptr<dwio::common::SeekableInputStream> stream = stripe.getStream(id, false);
  if (stream) {
    notNullDecoder_ = createBooleanRleDecoder(std::move(stream), encodingKey);
  }

  // We always initialize indexStream_ because indices are needed as
  // soon as there is a single filter that can trigger row group skips
  // anywhere in the reader tree. This is not known at construct time
  // because the first filter can come from a hash join or other run
  // time pushdown.
  id = kFormat == DwrfFormat::kDwrf ? encodingKey.forKind(proto::Stream_Kind_ROW_INDEX) : encodingKey.forKind(proto::orc::Stream_Kind_ROW_INDEX);
  indexStream_ = stripe.getStream(
      encodingKey.forKind(proto::orc::Stream_Kind_ROW_INDEX), false);
}
```

Reply:
OK, let‘s reduce the duplications in a new PR, I will ask my partner to do this for me in the following days.